keymap.c

```diff
@@ -1298,6 +1298,186 @@ enum CommandResult mutt_parse_bind(struct Buffer *buf, struct Buffer *s,
+ *
+ * Function iterate through Keymap and frees keys
+ * defined either by macro or bind
+ */
+static void km_unbind_all(struct Keymap **map, unsigned long mode)
+{
+  struct Keymap *cur, *next, *first, *last;
+
+  first = last = NULL;
+  cur = *map;
+
+  while (cur != NULL)
+  {
+    next = cur->next;
+    if ((mode & MUTT_UNBIND && cur->macro == NULL) ||
+        (mode & MUTT_UNMACRO && cur->macro != NULL))
+    {
+      FREE(&cur->macro);
+      FREE(&cur->keys);
+      FREE(&cur->desc);
+      FREE(&cur);
+    }
+    else if (first == NULL)
+    {
+      first = last = cur;
+    }
+    else if (last != NULL)
+    {
+      last->next = cur;
+      last = cur;
+    }
+    else
+    {
+      last = cur;
+    }
+    cur = next;
+  }
+
+  if (last != NULL)
+    last->next = NULL;
+
+  *map = first;
+}
+
+/**
+ * mutt_parse_unbind - Parse the 'unbind' command - Implements ::command_t
+ * @param buf
+ * @param s
+ * @param data
+ * @param err
+ *
+ * @return r
+ *
+ * Command unbinds specific binging in menu-name, in all menu-names
+ * or all bindings in all menu-names.
+ * unbind <menu-name[,…]|*> [<key_sequence>]
+ */
+enum CommandResult mutt_parse_unbind(struct Buffer *buf, struct Buffer *s,
+                                     unsigned long data, struct Buffer *err)
+{
+  int menu[sizeof(Menus) / sizeof(struct Mapping) - 1], nummenus;
+  enum CommandResult r = MUTT_CMD_SUCCESS;
+  bool all_menus = false, all_keys = false;
+  char *key = NULL;
+  char *cmd = "unbind";
+  if ((data & MUTT_UNMACRO) == MUTT_UNMACRO)
+    cmd = "unmacro";
+
+  mutt_extract_token(buf, s, 0);
+  if (mutt_str_strcmp(buf->data, "*") == 0) //when menu-names is * (all)
+    all_menus = true;
+  else // when menu-names is comma separated list
+    parse_menu(menu, buf->data, mutt_array_size(menu), &nummenus, err, cmd);
+
+  if (MoreArgs(s)) // second arg - key
+  {
+    mutt_extract_token(buf, s, 0);
+    key = buf->data;
+  }
+  else
+    all_keys = true;
+
+  if ((all_menus && mutt_str_strcmp(key, "q") == 0) ||
+      (all_menus && mutt_str_strcmp(key, ":") == 0))
+  {
+    mutt_buffer_printf(err, _("%s: cannot unbind '%s' from all menus"), cmd, key);
+    return MUTT_CMD_ERROR;
+  }
+
+  if (MoreArgs(s))
+  {
+    mutt_buffer_printf(err, _("%s: too many arguments"), cmd);
+    return MUTT_CMD_ERROR;
+  }
+  else // Here comes the logic, Mr. Tuvok.
+  {
+    if (all_menus == true && all_keys == true)
+    {
+      for (int i = 0; i < MENU_MAX; ++i)
+      {
+        km_unbind_all(&Keymaps[i],data);
+        if ((data & MUTT_UNBIND) == MUTT_UNBIND && i == MENU_GENERIC)
+        {
+          km_bindkey("<return>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey("<enter>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey(":", MENU_GENERIC, OP_ENTER_COMMAND);
+          km_bindkey("q", MENU_GENERIC, OP_QUIT);
+          km_bindkey("?", MENU_GENERIC, OP_HELP);
+        }
+      }
+    }
+    else if (all_menus == false && all_keys == true)
```

Comment:
Unless I'm mistaken, this and the above block can be merged into:

```C
if (all_keys)
{
  // Change the name of this one because I'm too lazy to think of a good one right now :)
  int upper_bound = all_menus ? MENU_MAX : nummenus;

  for (int i = 0; i < upper_bound; ++i)
  {
     km_unbind_all(&Keymaps[i],data);
     if ((data & MUTT_UNBIND) == MUTT_UNBIND && i == MENU_GENERIC)
     {
       km_bindkey("<return>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
       km_bindkey("<enter>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
       km_bindkey(":", MENU_GENERIC, OP_ENTER_COMMAND);
       km_bindkey("q", MENU_GENERIC, OP_QUIT);
       km_bindkey("?", MENU_GENERIC, OP_HELP);
     }
   }
}
```

Reply:
I thought I replied before to this too, but apparently not.
After merging those blocks, it behaves really weird. unbind and unmacro removes all keys (it's good to se with unmacro and macro command to show what has left. The keys in km_bindkeys() are not bound , hits the condition.
And then for example after `unbind index` trying to get into command prompt `:`  fills the whole command prompt from left to right with colons.

This could be probably the reason. 
`        km_unbind_all(&Keymaps[i],data);`
`        km_unbind_all(&Keymaps[menu[i]],data);`

Reply:
should be solved in the latest commit…